### ISSUE TITLE
`interp1d` behaviour inconsistent with Numpy's?

### ISSUE THREAD

**Issue by mathisc** [scipy/scipy#14531]
Scipy interpolator returns `nan` if one of the neighbor is a `nan` **even when interpolation is not actually needed**.
This looks inconsistent with Numpy's behaviour : 

```python
>>> import numpy as np
>>> from scipy.interpolate import interp1d
>>> np.interp([0,1,2], xp=[0,1,2], fp=[np.nan,0,1])
array([nan,  0.,  1.])
>>> interp1d([0, 1, 2], [np.nan, 0, 1])([0,1,2])
array([nan, nan,  1.])
```
Is that an expected behaviour ? 

Thanks a lot,

Version information : 
```python
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.6.0 1.19.4 sys.version_info(major=3, minor=8, micro=5, releaselevel='final', serial=0)
```

**Comment by ev-br** [comment]
In principle, we can review a PR which changes it for the "linear" kind. Might be a backcompat issue, too --- touching interp1d always breaks something for somebody (space bar overheating it is).

**Comment by ev-br** [comment]
Also note that if the `xp` array has explicit `dtype=float`, `np.interp1d(..., kind="linear")` delegates to `np.interp`, so that the behavior is what you expect: 
```
In [4]: interp1d([0., 1., 2.], [np.nan, 0, 1])([0,1,2])
Out[4]: array([nan,  0.,  1.])
```
So a simplest fix would be to pass integer dtypes to numpy here: https://github.com/scipy/scipy/blob/master/scipy/interpolate/interpolate.py#L527

**Comment by czgdp1807** [comment]
Can I try fixing this? @ev-br

**Comment by mathisc** [comment]
@ev-br : the issue is that I am not working with integers, I just used them just to reproduce the issue in the example I sent

**Comment by ev-br** [comment]
@czgdp1807 certainly! No need to ask permission BTW:-). 

@mathisc what is the dtype you're actually using it then? Like I said, interp1d tries to delegate to np.interp where it can

**Comment by czgdp1807** [comment]
I have a question. Why data type is one of the factors in deciding whether to delegate the interpolation task to numpy or not?

**Comment by ev-br** [comment]
Because np.interp does not handle eg complex valued `y`. (Or at least did not when this code was last looked at)

**Comment by czgdp1807** [comment]
I see. May be everything other than that (real inputs) can be delegated to numpy? I tried, the following test fails.

```python
============================================================= FAILURES ==============================================================
__________________________________________________ TestInterp1D.test_linear_dtypes __________________________________________________
scipy/interpolate/tests/test_interpolate.py:278: in test_linear_dtypes
    assert_equal(yp.dtype, dtyp)
E   AssertionError: 
E   Items are not equal:
E    ACTUAL: dtype('float64')
E    DESIRED: <class 'numpy.float16'>
        dtyp       = <class 'numpy.float16'>
        self       = <scipy.interpolate.tests.test_interpolate.TestInterp1D object at 0x15dd37700>
        x          = array([0., 1., 2., 3., 4., 5., 6., 7.], dtype=float16)
        y          = array([0., 1., 2., 3., 4., 5., 6., 7.], dtype=float16)
        yp         = array([0., 1., 2., 3., 4., 5., 6., 7.])
====================================================== short test summary info ======================================================
FAILED scipy/interpolate/tests/test_interpolate.py::TestInterp1D::test_linear_dtypes - AssertionError: 
================================= 1 failed, 611 passed, 1 skipped, 1 deselected, 1 xfailed in 9.26s =================================
False
```

**Comment by ev-br** [comment]
Yeah, this should keep backwards compatibility, including dtypes

**Comment by mathisc** [comment]
@ev-br : I was using float types. Every time you have an array that contains `np.nan` the dtype is required to be float : 
```python
>>> np.array([1,2,3, np.nan], dtype=int)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: cannot convert float NaN to integer
```

**Comment by ev-br** [comment]
Yes, the lack of integer nans in numpy is known. Then what exactly is the issue, can you add an actual reproducer if your OP using integers is not reflective of your actual issue. Better still, check if gh-14601 fixes the issue for you

**Comment by mathisc** [comment]
@ev-br : The issue I opened is correct. In my previous post, I was just underlining that this issue can only been seen with float types (as it requires an array containing np.nan). 
I'll try out the fix when I get a chance to. 

Thanks,

Claire

**Comment by ev-br** [comment]
Reopening until @mathisc confirms that gh-14601 fixes the original issue or comes up with a fix, as per https://github.com/scipy/scipy/issues/14531#issuecomment-902205813

**Comment by czgdp1807** [comment]
> Reopening until @mathisc confirms that [gh-14601](https://github.com/scipy/scipy/pull/14601) fixes the original issue or comes up with a fix, as per [#14531 (comment)](https://github.com/scipy/scipy/issues/14531#issuecomment-902205813)

Sure. Let me know if I can be of any help here. Thanks.

**Comment by ev-br** [comment]
Detaching the milestone, this is not releasr-critical.If a fix arrives, we'll review it though.

**Comment by ev-br** [comment]
No follow-up for several months, let's close the issue. Thanks all
